gnrc: networking minimal

OlegHahm opened the pull request:
This is a minimalistic example for RIOT's gnrc network stack. It just boots and waits to be pinged. Requires less than 10kB of RAM and less than 30kB of ROM on a ARM Cortex-M platform.

~~Depends on #4282.~~

cgundogan commented on examples/gnrc_minimal/main.c:
```diff
@@ -0,0 +1,49 @@
+    size_t numof = gnrc_netif_get(ifs);
+    if (numof > 0) {
+        gnrc_ipv6_netif_t *entry = gnrc_ipv6_netif_get(ifs[0]);
+        for (int i = 0; i < GNRC_IPV6_NETIF_ADDR_NUMOF; i++) {
```
you may also want to reduce `GNRC_IPV6_NETIF_ADDR_NUMOF` to something around `4` for
```
ff02::1/128
ff02::1:XXX/64
ff02::2/128
fe80::XXX/64
```

I think the default value is around 6-7 and each `gnrc_ipv6_netif_addr_t` struct is in the range of ~54 bytes

OlegHahm replied:
Nice idea. I was actually too lazy to count what the minimum number would be. ;)

cgundogan replied:
I just looked at the ifconfig output of a freshly rebooted node (:

OlegHahm commented:
Integrated @cgundogan's proposal to save some additional 100 bytes.

LudwigKnuepfer commented:
Dependency has been merged, waiting for anything else?

OlegHahm commented:
Not that I know of - should I squash?

cgundogan commented:
Yes, squash please.

OlegHahm commented:
squashed

PeterKietzmann commented:
`examples/gnrc_minimal/main.c:30: style (variableScope): The scope of the variable 'ipv6_addr' can be reduced`

emmanuelsearch commented:
For some weird reason, on this branch, gnrc_networking does not compile. I get the following error (on my Linux VM). Any idea where this comes from?

```
/home/emmanuel/Bureau/RIOT/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c: In function '_event_loop':
/home/emmanuel/Bureau/RIOT/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c:255:17: error: implicit declaration of function 'gnrc_sixlowpan_nd_router_abr_remove' [-Werror=implicit-function-declaration]
                 gnrc_sixlowpan_nd_router_abr_remove(
                 ^
/home/emmanuel/Bureau/RIOT/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c:256:26: error: 'gnrc_sixlowpan_nd_router_abr_t' undeclared (first use in this function)
                         (gnrc_sixlowpan_nd_router_abr_t *)msg.content.ptr);
                          ^
/home/emmanuel/Bureau/RIOT/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c:256:26: note: each undeclared identifier is reported only once for each function it appears in
/home/emmanuel/Bureau/RIOT/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c:256:58: error: expected expression before ')' token
                         (gnrc_sixlowpan_nd_router_abr_t *)msg.content.ptr);
                                                          ^
/home/emmanuel/Bureau/RIOT/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c:260:17: error: implicit declaration of function 'gnrc_sixlowpan_nd_router_gc_nc' [-Werror=implicit-function-declaration]
                 gnrc_sixlowpan_nd_router_gc_nc((gnrc_ipv6_nc_t *)msg.content.ptr);
                 ^
cc1: all warnings being treated as errors
make[4]: *** [/home/emmanuel/Bureau/RIOT/examples/gnrc_networking/bin/iotlab-m3/gnrc_ipv6/gnrc_ipv6.o] Erreur 1
make[3]: *** [ALL--/home/emmanuel/Bureau/RIOT/sys/net/gnrc/network_layer/ipv6] Erreur 2
make[2]: *** [ALL--/home/emmanuel/Bureau/RIOT/sys/net/gnrc] Erreur 2
make[1]: *** [ALL--/home/emmanuel/Bureau/RIOT/sys] Erreur 2
make: *** [all] Erreur 2
```

emmanuelsearch commented:
On Mac OS X the error thrown (when trying to compile gnrc_networking) becomes:

```
/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/net/network_layer/fib/fib.c:1286:25: error: equality comparison with extraneous parentheses [-Werror,-Wparentheses-equality]
            if((*fib_sr == &table->data.source_routes->headers[i])) {
                ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/net/network_layer/fib/fib.c:1286:25: note: remove extraneous parentheses around the comparison to silence this warning
            if((*fib_sr == &table->data.source_routes->headers[i])) {
               ~        ^                                        ~
/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/net/network_layer/fib/fib.c:1286:25: note: use '=' to turn this equality comparison into an assignment
            if((*fib_sr == &table->data.source_routes->headers[i])) {
                        ^~
                        =
1 error generated.
make[3]: *** [/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/examples/gnrc_networking/bin/native/fib/fib.o] Error 1
make[2]: *** [ALL--/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/net/network_layer/fib] Error 2
make[1]: *** [ALL--/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys] Error 2
make: *** [all] Error 2
```

emmanuelsearch commented:
And one more: if I try to compile gnrc_minimal on OS X for boards other than native, I get the following error complaining about ports:

```
$ make clean all BOARD=iotlab-m3
ls: /dev/tty.usbserial*: No such file or directory
Warning: no PORT set!
ls: /dev/tty.usbserial*: No such file or directory
ls: /dev/tty.usbserial*: No such file or directory
ls: /dev/tty.usbserial*: No such file or directory
ls: /dev/tty.usbserial*: No such file or directory
ls: /dev/tty.usbserial*: No such file or directory
ls: /dev/tty.usbserial*: No such file or directory
Building application "gnrc_minimal" for "iotlab-m3" with MCU "stm32f1".

ls: /dev/tty.usbserial*: No such file or directory
ls: /dev/tty.usbserial*: No such file or directory
In file included from /Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/sys/include/net/ipv6/addr.h:32:0,
                 from /Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/examples/gnrc_minimal/main.c:24:
/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/core/include/byteorder.h: In function ‘byteorder_swaps’:
/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/core/include/byteorder.h:288:5: erreur: implicit declaration of function ‘__builtin_bswap16’ [-Werror=implicit-function-declaration]
cc1: all warnings being treated as errors

make[1]: *** [/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/examples/gnrc_minimal/bin/iotlab-m3/gnrc_minimal/main.o] Error 1
make: *** [all] Error 2
```

OlegHahm commented:
* rebased on master
* use minimal standard PRNG to save more RAM (thanks to @kaspar030 for implementing and @gebart for reviewing)
* addressed @PeterKietzmann's comment to reduce the scope of a variable

@emmanuelsearch, are you sure that these errors do not occur on current master?

OlegHahm commented:
Btw. we're down to 26,5kB ROM and 6,8kB RAM on iotlab-m3.

kaspar030 commented:
On iot-lab, prng_musl_lcg might some more bytes.

emmanuelsearch commented:
No, I do not get errors compiling gnrc_networking on current master for native on OS X, or on native, iotlab-m3 or samr21-xpro on my Linux VM.

OlegHahm commented:
That's very weird since there are no changes outside the `gnrc_minimal` folder. I have no clue for now.

emmanuelsearch commented:
By the way, here are the numbers I have when I compile the current gnrc_minimal for iotlab-m3 on my Linux VM:

```
   text       data        bss        dec        hex    filename
  45764       2236       6788      54788       d604    /home/emmanuel/Bureau/RIOT/examples/gnrc_minimal/bin/iotlab-m3/gnrc_minimal.elf
```

OlegHahm commented:
@kaspar030, musl_lcq saves 38 bytes ROM and no RAM on my machine.

@emmanuelsearch, strange. I gonna try on another machine with a different toolchain.

haukepetersen commented:
Just tried it on my office machine (`arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 4.8.4 20140725 (release) [ARM/embedded-4_8-branch revision 213147]`):
```
   text	   data	    bss	    dec	    hex	filename
  26624	    132	   6748	  33504	   82e0	/home/hauke/dev/riot/RIOT/examples/gnrc_minimal/bin/iotlab-m3/gnrc_minimal.elf
```

haukepetersen commented:
@emmanuelsearch: Seems to me, like your toolchain is not using the newlib nano specs. You can check this, by putting a quick output into the `boards/Makefile.include.cortexm_common` Makefile. Just change the second last paragraph to something like:
```
# use the nano-specs of Newlib when available
ifeq ($(shell $(LINK) -specs=nano.specs -E - 2>/dev/null >/dev/null </dev/null ; echo $$?),0)
export LINKFLAGS += -specs=nano.specs -lc -lnosys
$(info USING NANO SPECS)
else
$(info NOT USING NANO SPECS)
endif
```

emmanuelsearch commented:
You guessed correctly: the nano specs were not being used.

haukepetersen commented:
solved :-)

OlegHahm commented:
Should I squash?

@kaspar030, should I change to musl lcq or stay with minimal standard PRNG? How should musl's PRNG behave on 8 and 16 bit platforms?

emmanuelsearch commented:
looks much better with nano specs indeed

```
   text       data        bss        dec        hex    filename
  26520        132       6748      33400       8278    /home/emmanuel/Bureau/RIOT/examples/gnrc_minimal/bin/iotlab-m3/gnrc_minimal.elf
```

OlegHahm commented:
That looks quite familiar to me. :)

haukepetersen commented:
Hm, would be very interesting to compile with different releases of the arm toolchain and compare the results...

OlegHahm commented:
@emmanuelsearch, @haukepetersen, @authmillenon, are you okay with the README? Are we ready to go?

authmillenon commented on examples/gnrc_minimal/README.md:
```diff
@@ -0,0 +1,27 @@
+ * Reducing the packet buffer size from 6kB to 512 bytes.
+ * Reducing the maximum neighbor cache size from 8 to 1.
+
+Please take a look at the Makefile to see how the configuration is done.
```
Though I get what you mean (look at the comments in the Makefile) this is exactly what our doc fixes are meant to minimize: the need to look at code (yes, I count the Makefile as code now, deal with it ^^). For this iteration it is okay I guess, but (optionally) it would be IMHO better to explain how this application achieves the points stated above.

OlegHahm replied:
Still, this is not the task of an example's README.

authmillenon commented on examples/gnrc_minimal/Makefile:
```diff
@@ -0,0 +1,35 @@
+CFLAGS += -DGNRC_IPV6_NETIF_ADDR_NUMOF=4
+# Reduce sizes for GNRC's packet buffer and neighbor cache
+ifneq (,$(filter native,$(BOARD)))
+    CFLAGS += -DGNRC_PKTBUF_SIZE=2048 -DGNRC_IPV6_NC_SIZE=1
```
Maybe explained in the discussion above, but why is native different?

OlegHahm replied:
added a comment

OlegHahm commented:
IMO ready to be squashed and merged

OlegHahm commented:
@haukepetersen, @authmillenon, @emmanuelsearch, ping

emmanuelsearch commented:
tested again on native (Linux Mint VM) works.
ACK as soon as someone confirms this is working on hardware.

OlegHahm commented:
Squashed.

cgundogan commented:
I flashed this example on 4 iotlab-m3 nodes on the iotlab-testbed and a 5. node with the `gnrc_networking` example. I was not able to unicast/multicast ping6 the nodes that were running the minimal example

cgundogan commented:
pinging works on native, though

OlegHahm commented:
Strange but true.

OlegHahm commented:
Ah, no, not strange at all. It's a bug, not a feature: gnrc_minimal is built without IPHC, so you have disable this on gnrc_networking (`ifconfig 7 -iphc`). Will add a note to the README.

OlegHahm commented:
Updated the README.

cgundogan commented:
>...It's a bug, not a feature:...

Didn't you mean the other way around? (:
I will test with iphc disabled later in the evening and keep you updated.

OlegHahm commented:
I don't know what you're talking about. ;)

cgundogan commented:
tested again with iphc disabled. Looks good. ACK. Please squash.

OlegHahm commented:
squashed
